### PR TITLE
Display dialog when we detect a downgrade

### PR DESCRIPTION
**Pull request by erikjv**

### CODE REVIEW

**Comment by erikjv** [comment]
@TheOneRing can you do a first review and let me know if this is what we want?

**Comment by ownclouders** [comment]
Results for <strong>GUI-tests-ocis</strong> https://drone.owncloud.com/owncloud/client/15848/6/1

:boom: The GUI tests failed.

GUI Logs: https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/index.html

<details>

Server Logs: https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/serverlog.log

Screenshots:
- [Add_space_manually_from_sync_connection_window.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/Add_space_manually_from_sync_connection_window.png)
- [Adding_account_with_self_signed_certificate_for_the_first_time.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/Adding_account_with_self_signed_certificate_for_the_first_time.png)
- [Adding_account_with_vfs_enabled.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/Adding_account_with_vfs_enabled.png)
- [Adding_multiple_accounts.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/Adding_multiple_accounts.png)
- [Adding_normal_Account.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/Adding_normal_Account.png)
- [Both_original_and_copied_folders_can_be_synced.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/Both_original_and_copied_folders_can_be_synced.png)
- [Check_default_options_in_advanced_configuration.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/Check_default_options_in_advanced_configuration.png)
- [Delete_a_file.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/Delete_a_file.png)
- [Delete_a_file_and_a_folder.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/Delete_a_file_and_a_folder.png)
- [Delete_a_folder.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/Delete_a_folder.png)
- [Disable_VFS.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/Disable_VFS.png)
- [Enable_VFS.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/Enable_VFS.png)
- [File_with_long_name_can_be_synced.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/File_with_long_name_can_be_synced.png)
- [File_with_spaces_in_the_name_can_sync.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/File_with_spaces_in_the_name_can_sync.png)
- [Filenames_that_are_rejected_by_the_server_are_reported.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/Filenames_that_are_rejected_by_the_server_are_reported.png)
- [Invalid_system_names_are_synced_in_linux.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/Invalid_system_names_are_synced_in_linux.png)
- [Many_subfolders_can_be_synced.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/Many_subfolders_can_be_synced.png)
- [Modify_orignal_content_of_a_file_with_special_character.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/Modify_orignal_content_of_a_file_with_special_character.png)
- [Move_folder_and_file_from_level_5_sub-folder_to_sync_root.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/Move_folder_and_file_from_level_5_sub-folder_to_sync_root.png)
- [Move_two_folders_and_a_file_down_to_the_level_5_sub-folder.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/Move_two_folders_and_a_file_down_to_the_level_5_sub-folder.png)
- [Rename_a_file_and_a_folder.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/Rename_a_file_and_a_folder.png)
- [Skip_sync_folder_configuration.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/Skip_sync_folder_configuration.png)
- [Syncing_a_file_from_the_server_and_creating_a_conflict.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/Syncing_a_file_from_the_server_and_creating_a_conflict.png)
- [Syncing_a_file_to_the_server.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/Syncing_a_file_to_the_server.png)
- [Syncing_a_folder_to_the_server.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/Syncing_a_folder_to_the_server.png)
- [Syncing_all_files_and_folders_from_the_server.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/Syncing_all_files_and_folders_from_the_server.png)
- [Syncing_file_of_1_GB_size.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/Syncing_file_of_1_GB_size.png)
- [Syncing_folders_each_having_500_files.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/Syncing_folders_each_having_500_files.png)
- [Tabs_in_toolbar_looks_correct.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/Tabs_in_toolbar_looks_correct.png)
- [Try_to_enable_experimental_vfs_option_and_cancel_it.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/Try_to_enable_experimental_vfs_option_and_cancel_it.png)
- [User_with_Editor_role_can_edit_the_file.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/User_with_Editor_role_can_edit_the_file.png)
- [User_with_Manager_role_can_add_files_and_folders.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/User_with_Manager_role_can_add_files_and_folders.png)
- [User_with_Viewer_role_can_open_the_file.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/User_with_Viewer_role_can_open_the_file.png)
- [User_with_Viewer_role_cannot_edit_the_file.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/User_with_Viewer_role_cannot_edit_the_file.png)
- [VFS_is_disabled_by_default.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/VFS_is_disabled_by_default.png)
- [Verify_one_empty_folder_with_a_length_longer_than_the_allowed_limit_will_not_be_synced.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/Verify_one_empty_folder_with_a_length_longer_than_the_allowed_limit_will_not_be_synced.png)
- [Verify_pre_existing_folders_in_local_(Desktop_client)_are_copied_over_to_the_server.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/Verify_pre_existing_folders_in_local_(Desktop_client)_are_copied_over_to_the_server.png)
- [Verify_that_you_can_create_a_subfolder_with_long_name.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/Verify_that_you_can_create_a_subfolder_with_long_name.png)
- [extract_a_zip_file_in_the_sync_folder.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/extract_a_zip_file_in_the_sync_folder.png)
- [logging_out.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/logging_out.png)
- [login_after_loggin_out.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/login_after_loggin_out.png)
- [remove_an_account_connection.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/remove_an_account_connection.png)
- [remove_the_only_account_connection.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/remove_the_only_account_connection.png)
- [various_types_of_files_can_be_synced_from_client_to_server.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/various_types_of_files_can_be_synced_from_client_to_server.png)
- [various_types_of_files_can_be_synced_from_server_to_client.png](https://cache.owncloud.com/public/owncloud/client/15848/ocis/guiReportUpload/screenshots/various_types_of_files_can_be_synced_from_server_to_client.png)

</details>

File: src/gui/accountmanager.cpp [hunk 1]
```diff
@@ -150,7 +150,7 @@ void AccountManager::backwardMigrationSettingsKeys(QStringList *deleteKeys, QStr
         const auto &childGroups = settings->childGroups();
         for (const auto &accountId : childGroups) {
             settings->beginGroup(accountId);
-            const int accountVersion = settings->value(versionC(), 1).toInt();
+            const int accountVersion = settings->value(versionC(), 13).toInt(); // See https://github.com/owncloud/client/issues/10473
```
Comment (TheOneRing) [inline src/gui/accountmanager.cpp:153]: I guess a short comment in addition to the url won't hurt

File: src/gui/application.cpp [hunk 2]
```diff
@@ -93,32 +93,25 @@ bool Application::configVersionMigration()
...
             return false;
         }
 
         auto settings = ConfigFile::settingsWithGroup(QStringLiteral("foo"));
         settings->endGroup();
 
         // Wipe confusing keys from the future, ignore the others
-        for (const auto &badKey : qAsConst(deleteKeys))
+        for (const auto &badKey : qAsConst(deleteKeys)) {
             settings->remove(badKey);
```
Comment (TheOneRing) [inline src/gui/application.cpp:113]: I thought we no longer remove random settings but prevent loading the old config?

File: src/gui/application.cpp [hunk 3]
```diff
@@ -93,32 +93,25 @@ bool Application::configVersionMigration()
...
+                             "%1<br>"
+                             "<br>"
+                             "The current configuration file was already backed up to <i>%2</i>.")
+                              .arg(boldMessage, backupFile);
+
+        if (!askContinueToDowngrade(message)) {
             return false;
         }
 
         auto settings = ConfigFile::settingsWithGroup(QStringLiteral("foo"));
```
[the remark this answers is not in the document]
Comment (erikjv) [reply src/gui/application.cpp:108]: I'll comment on this

[the remark this answers is not in the document]
Comment (erikjv) [reply src/gui/application.cpp]: done, removed it.

File: src/gui/application.cpp (hunk 2 above)
Comment (erikjv) [reply src/gui/application.cpp]: removed it.

File: src/gui/accountmanager.cpp [hunk 4]
```diff
@@ -82,8 +82,10 @@ auto capabilitesC()
 }
 
 // The maximum versions that this client can read
-static const int maxAccountsVersion = 2;
-static const int maxAccountVersion = 1;
+// We don't use these versions anymore, see https://github.com/owncloud/client/issues/10473 . These
+// values are only written, and that prevents older clients from loading these newer settings.
+static const int maxAccountsVersion = 13;
```
Comment (TheOneRing) [inline src/gui/accountmanager.cpp:87]: Off topic but the static isn't required and we could use constexpr.

File: src/gui/application.cpp [hunk 5]
```diff
@@ -67,60 +67,32 @@ Q_LOGGING_CATEGORY(lcApplication, "gui.application", QtInfoMsg)
...
-        if (!deleteKeys.isEmpty()) {
-            boldMessage = tr("Continuing will mean <b>deleting these settings</b>.");
-        } else {
-            boldMessage = tr("Continuing will mean <b>ignoring these settings</b>.");
-        }
-
-        QMessageBox box(
-            QMessageBox::Warning,
-            Theme::instance()->appNameGUI(),
+    if (clientVersion > configVersion) {
```
Comment (TheOneRing) [inline src/gui/application.cpp:82]: I guess we should limit this to the major version.
Currently we even include the build number.

File: src/gui/application.cpp [hunk 6]
```diff
@@ -67,60 +67,32 @@ Q_LOGGING_CATEGORY(lcApplication, "gui.application", QtInfoMsg)
...
-
-        // Wipe confusing keys from the future, ignore the others
-        for (const auto &badKey : qAsConst(deleteKeys))
-            settings->remove(badKey);
+        QTimer::singleShot(0, qApp, &QApplication::quit);
+        return false;
     }
 
+
+    // Upgrade settings
```
Comment (TheOneRing) [inline src/gui/application.cpp:94]: misleading comment

File: src/gui/folder.cpp [hunk 7]
```diff
@@ -81,7 +81,9 @@ auto priorityC()
     return QStringLiteral("priority");
 }
 
-constexpr int SettingsVersionC = 5;
+// We don't use these versions anymore, see https://github.com/owncloud/client/issues/10473 . This
+// value is only written, and that prevents older clients from loading these newer settings.
+constexpr int SettingsVersionC = 13;
```
Comment (TheOneRing) [inline src/gui/folder.cpp:86]: Hmmm could it make sense to move all those 13 to ConfigFile::UnusedLegacySettingsVersionNumber ?

**Review by TheOneRing — CHANGES_REQUESTED** [review]

**Comment by TheOneRing** [comment]
Can you by now explain what MultiFolder is? https://github.com/owncloud/client/issues/9117

File: src/gui/application.cpp (hunk 5 above)
Comment (erikjv) [reply src/gui/application.cpp:82]: done

File: src/gui/application.cpp (hunk 6 above)
Comment (erikjv) [reply src/gui/application.cpp:94]: clarified the comment

File: src/gui/folder.cpp (hunk 7 above)
Comment (erikjv) [reply src/gui/folder.cpp:86]: done

**Comment by erikjv** [comment]
> Can you by now explain what MultiFolder is? #9117

Not yet, have to dive a little bit deeper.

File: src/gui/application.cpp [hunk 8]
```diff
@@ -67,60 +67,32 @@ Q_LOGGING_CATEGORY(lcApplication, "gui.application", QtInfoMsg)
...
-        if (!deleteKeys.isEmpty()) {
-            boldMessage = tr("Continuing will mean <b>deleting these settings</b>.");
-        } else {
-            boldMessage = tr("Continuing will mean <b>ignoring these settings</b>.");
-        }
-
-        QMessageBox box(
-            QMessageBox::Warning,
-            Theme::instance()->appNameGUI(),
+    if (clientVersion.majorVersion() > configVersion.majorVersion()) {
```
Comment (TheOneRing) [inline src/gui/application.cpp:82]: Appears to be inverted

Comment (erikjv) [reply src/gui/application.cpp]: done

**Comment by TheOneRing** [comment]
Rest looks good.

File: src/gui/application.cpp [hunk 9]
```diff
@@ -67,60 +67,32 @@ Q_LOGGING_CATEGORY(lcApplication, "gui.application", QtInfoMsg)
...
-            return false;
-        }
-
-        auto settings = ConfigFile::settingsWithGroup(QStringLiteral("foo"));
-        settings->endGroup();
-
-        // Wipe confusing keys from the future, ignore the others
-        for (const auto &badKey : qAsConst(deleteKeys))
-            settings->remove(badKey);
+        QTimer::singleShot(0, qApp, &QApplication::quit);
```
[the remark this answers is not in the document]
Comment (erikjv) [reply src/gui/application.cpp:89]: done

[the remark this answers is not in the document]
Comment (erikjv) [reply src/gui/application.cpp:89]: @TheOneRing please check my solution, the language part isn't set up yet, so I don't know how this will work with translations.

File: src/gui/main.cpp [hunk 10]
```diff
@@ -261,6 +261,12 @@ int main(int argc, char **argv)
 
         platform->setApplication(&app);
 
+        // Check if the user upgraded or downgraded. We do this as early as possible, to detect
+        // a possible downgrade.
+        if (!OCC::Application::configVersionMigration()) {
```
Comment (TheOneRing) [inline src/gui/main.cpp:266]: Hmm line 260 already accesses the cfg file.
Can we move the check even before that? I guess we have to parse the options for --confdir. (Off topic: we should probably not execute unrelated code in parseOptions).

I guess we can then move the whole function here, and give it a name that represents its current purpose.
(We will no longer need the qtimer connected to quit)

Comment (erikjv) [reply src/gui/main.cpp]: done

File: src/gui/main.cpp [hunk 11]
```diff
@@ -207,6 +207,43 @@ CommandLineOptions parseOptions(const QStringList &arguments)
     return out;
 }
 
+/**
+ * Maybe a newer version of the client was used with this config file:
+ * if so, backup, confirm with user and remove the config that can't be read.
+ */
+bool configVersionMigration()
```
Comment (TheOneRing) [inline src/gui/main.cpp:214]: we backup and don't migrate? So I guess we should update the name.

Comment (erikjv) [reply src/gui/main.cpp:216]: changed the comment: only migrate when updating

Comment (erikjv) [reply src/gui/main.cpp]: renamed it to `checkClientVersion()`
